### PR TITLE
Update `GradientDescentTrainer` to automatically create directory for `serialization_dir`

### PR DESCRIPTION
**Pull request by himkt**
Background: https://github.com/allenai/allennlp/issues/4939

This PR makes `GradientDescentTrainer` create `serialization_dir` itself.

@epwalsh Please take a look.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -91,6 +91,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Torch version bumped to 1.7.1 in Docker images.
 - 'master' branch renamed to 'main'
 - Torch version bumped to 1.7.1 in Docker images.
+- GradientDescentTrainer makes `serialization_dir` in its instantiated step.
```
Comment (epwalsh) [inline CHANGELOG.md:94]:
```suggestion
- GradientDescentTrainer makes `serialization_dir` when it's instantiated, if it doesn't exist.
```

Comment (epwalsh) [reply CHANGELOG.md:94]: Could you move this up to the "Unreleased" section?

**Comment by himkt** [comment]
@epwalsh Thank you for the quick review. I revised the PR. PTAL.

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+- GradientDescentTrainer makes `serialization_dir` when it's instantiated, if it doesn't exist.
```
Comment (epwalsh) [inline CHANGELOG.md:10]:
```suggestion
### Changed

- GradientDescentTrainer makes `serialization_dir` when it's instantiated, if it doesn't exist.
```

File: allennlp/training/trainer.py [hunk 3]
```diff
@@ -514,6 +514,7 @@ def __init__(
 
         self._checkpointer: Optional[Checkpointer] = checkpointer
         if checkpointer is None and serialization_dir is not None:
+            os.makedirs(serialization_dir, exist_ok=True)
```
Comment (epwalsh) [inline allennlp/training/trainer.py:517]: So if `checkpointer` is not `None` we might still not create the serialization directory? I think we want to create it either way, right?

**Review by epwalsh — CHANGES_REQUESTED** [review]

File: allennlp/training/trainer.py [hunk 4]
```diff
@@ -81,6 +81,10 @@ def __init__(
         self._rank = local_rank
         self._primary = self._rank == 0
         self._world_size = world_size
+
```
Comment (epwalsh) [inline allennlp/training/trainer.py:84]:
```suggestion

```

**Review by epwalsh — APPROVED** [review]
Thanks @himkt! I just made one little change so that the serialization directory is actually created in the base `Trainer` class.

**Pull request merged**